man/man3/TelodendriaConfig.3
.Dd $Mdocdate: December 10 2022 $
.Dt TELODENDRIACONFIG 3
.Os Telodendria Project
.Sh NAME
.Nm TelodendriaConfig
.Nd Parse the configuration file into a structure.
.Sh SYNOPSIS
.In TelodendriaConfig.h
.Ft TelodendriaConfig *
.Fn TelodendriaConfigParse "HashMap *" "LogConfig *"
.Ft void
.Fn TelodendriaConfigFree "TelodendriaConfig *"
.Sh DESCRIPTION
.Pp
Validate and maintain the Telodendria server's configuration data. This API
builds on the JSON API to add Telodendria-specific parsing. It takes a
fully-parsed JSON object and converts it into a TelodendriaConfig, which is
much more structured and easier to work with than the JSON. The config
structure is not opaque like many other structures in Telodendria. This is
intentional; defining functions for all of the fields would just add a lot
of unecessary overhead. The structure is defined as follows:
.Bd -literal -offset indent
typedef struct TelodendriaConfig
{
	char *serverName;
	char *baseUrl;
	char *identityServer;

	char *uid;
	char *gid;
	char *dataDir;

	unsigned short listenPort;
	unsigned int flags;
	unsigned int threads;
	unsigned int maxConnections;

	size_t maxCache;

	char *logTimestamp;
	int logLevel;
} TelodendriaConfig;
.Ed
.Pp
Since the configuration will live in memory for a long time, it is important
that unused values are freed as soon as possible. Therefore, the Telodendria
structure is not opaque; values are accessed directly, and they can be
freed as the program wishes. Do note that if you're going to free a value, you
should set it to NULL, because
.Fn TelodendriaConfigFree
will unconditionally call
.Fn Free
on all values.
.Pp
The flags variable in this structure is a bit field that contains the OR-ed values
of any of the given flags:
.Bd -literal -offset indent
typedef enum TelodendriaConfigFlag
{
	TELODENDRIA_FEDERATION,
	TELODENDRIA_REGISTRATION,

	TELODENDRIA_LOG_COLOR,

	TELODENDRIA_LOG_FILE,
	TELODENDRIA_LOG_STDOUT,
	TELODENDRIA_LOG_SYSLOG
} TelodendriaConfigFlag;
.Ed
.Pp
Do note that the actual values of these enums are omitted, but they can be
OR-ed together and added to flags.
.Pp
.Fn TelodendriaConfigParse
parses a JSON map, extracting the necessary values, validating them, and then
adding them to a new TelodendriaConfig for future use by the program. All values
are copied, so the JSON hash map can be safely freed if this function
succeeds. It takes a working log configuration so that messages can be written
to the log as the parsing progresses, to warn users about default values and
report errors, for example.
.Pp
.Fn TelodendriaConfigFree
frees all of the memory allocated for the given configuration. This function
unconditionally calls
.Fn Free
on all items in the structure, so make sure that items that were already freed
are NULL.
.Sh RETURN VALUES
.Pp
.Fn TelodendriaConfigParse
returns a TelodendriaConfig that is completely independent of the passed
configuration hash map, or NULL if one or more required values is missing, or
there was some other error while parsing the configuration.
.Sh SEE ALSO
.Xr Json 3


src/TelodendriaConfig.c
/*
 * Copyright (C) 2022 Jordan Bancino <@jordan:bancino.net>
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation files
 * (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
#include <TelodendriaConfig.h>
#include <Memory.h>
#include <Json.h>
#include <HashMap.h>
#include <Log.h>
#include <Array.h>
#include <Util.h>
#include <Db.h>

#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define CONFIG_REQUIRE(key, type) \
    value = HashMapGet(config, key); \
    if (!value) \
    { \
        Log(lc, LOG_ERR, "Missing required " key " directive."); \
        goto error; \
    } \
    if (JsonValueType(value) == JSON_NULL) \
    { \
        Log(lc, LOG_ERR, "Missing value for " key " directive."); \
        goto error; \
    } \
    if (JsonValueType(value) != type) \
    { \
        Log(lc, LOG_ERR, "Expected " key " to be of type " #type); \
        goto error; \
    }

#define CONFIG_COPY_STRING(into) \
    into = UtilStringDuplicate(JsonValueAsString(value));

#define CONFIG_OPTIONAL_STRING(into, key, default) \
    value = HashMapGet(config, key); \
    if (value && JsonValueType(value) != JSON_NULL) \
    { \
        if (JsonValueType(value) != JSON_STRING) \
        { \
            Log(lc, LOG_ERR, "Expected " key " to be of type JSON_STRING"); \
            goto error; \
        } \
        into = UtilStringDuplicate(JsonValueAsString(value)); \
    } \
    else \
    { \
        Log(lc, LOG_INFO, "Using default value " #default " for " key "."); \
        into = default ? UtilStringDuplicate(default) : NULL; \
    }

#define CONFIG_OPTIONAL_INTEGER(into, key, default) \
    value = HashMapGet(config, key); \
    if (value && JsonValueType(value) != JSON_NULL) \
    { \
        if (JsonValueType(value) != JSON_INTEGER) \
        { \
            Log(lc, LOG_ERR, "Expected " key " to be of type JSON_INTEGER"); \
            goto error; \
        } \
        into = JsonValueAsInteger(value); \
    } \
    else \
    { \
        Log(lc, LOG_INFO, "Using default value " #default " for " key "."); \
        into = default; \
    }

int
ConfigParseRunAs(LogConfig * lc, TelodendriaConfig * tConfig, HashMap * config)
{
    JsonValue *value;

    CONFIG_REQUIRE("uid", JSON_STRING);
    CONFIG_COPY_STRING(tConfig->uid);

    CONFIG_OPTIONAL_STRING(tConfig->gid, "gid", tConfig->uid);

    return 1;

error:
    return 0;
}

int
ConfigParseLog(LogConfig * lc, TelodendriaConfig * tConfig, HashMap * config)
{
    JsonValue *value;
    char *str;

    CONFIG_REQUIRE("output", JSON_STRING);
    str = JsonValueAsString(value);

    if (strcmp(str, "stdout") == 0)
    {
        tConfig->flags |= TELODENDRIA_LOG_STDOUT;
    }
    else if (strcmp(str, "file") == 0)
    {
        tConfig->flags |= TELODENDRIA_LOG_FILE;
    }
    else if (strcmp(str, "syslog") == 0)
    {
        tConfig->flags |= TELODENDRIA_LOG_SYSLOG;
    }
    else
    {
        Log(lc, LOG_ERR, "Invalid value for log.output: '%s'.", str);
        goto error;
    }

    CONFIG_OPTIONAL_STRING(str, "level", "message");

    if (strcmp(str, "message") == 0)
    {
        tConfig->logLevel = LOG_INFO;
    }
    else if (strcmp(str, "debug") == 0)
    {
        tConfig->logLevel = LOG_DEBUG;
    }
    else if (strcmp(str, "notice") == 0)
    {
        tConfig->logLevel = LOG_NOTICE;
    }
    else if (strcmp(str, "warning") == 0)
    {
        tConfig->logLevel = LOG_WARNING;
    }
    else if (strcmp(str, "error") == 0)
    {
        tConfig->logLevel = LOG_ERR;
    }
    else
    {
        Log(lc, LOG_ERR, "Invalid value for log.level: '%s'.", tConfig->logLevel);
        goto error;
    }

    Free(str);

    CONFIG_OPTIONAL_STRING(tConfig->logTimestamp, "timestampFormat", "default");

    if (strcmp(tConfig->logTimestamp, "none") == 0)
    {
        Free(tConfig->logTimestamp);
        tConfig->logTimestamp = NULL;
    }

    value = HashMapGet(config, "color");
    if (value && JsonValueType(value) != JSON_NULL)
    {
        if (JsonValueType(value) != JSON_BOOLEAN)
        {
            Log(lc, LOG_ERR, "Expected type JSON_BOOLEAN for log.color.");
            goto error;
        }

        if (JsonValueAsBoolean(value))
        {
            tConfig->flags |= TELODENDRIA_LOG_COLOR;
        }
    }

    return 1;

error:
    return 0;
}

TelodendriaConfig *
TelodendriaConfigParse(HashMap * config, LogConfig * lc)
{
    TelodendriaConfig *tConfig;
    JsonValue *value;

    if (!config || !lc)
    {
        return NULL;
    }

    tConfig = Malloc(sizeof(TelodendriaConfig));
    if (!tConfig)
    {
        return NULL;
    }

    memset(tConfig, 0, sizeof(TelodendriaConfig));

    CONFIG_OPTIONAL_INTEGER(tConfig->listenPort, "listen", 8008);

    CONFIG_REQUIRE("serverName", JSON_STRING);
    CONFIG_COPY_STRING(tConfig->serverName);

    value = HashMapGet(config, "baseUrl");
    if (value)
    {
        CONFIG_COPY_STRING(tConfig->baseUrl);
    }
    else
    {
        Log(lc, LOG_WARNING, "Base URL not specified. Assuming it's 'https://%s'.", tConfig->serverName);
        tConfig->baseUrl = Malloc(strlen(tConfig->serverName) + 10);
        if (!tConfig->baseUrl)
        {
            Log(lc, LOG_ERR, "Error allocating memory for default config value 'baseUrl'.");
            goto error;
        }

        sprintf(tConfig->baseUrl, "https://%s", tConfig->serverName);
    }

    CONFIG_OPTIONAL_STRING(tConfig->identityServer, "identityServer", NULL);

    value = HashMapGet(config, "runAs");
    if (value && JsonValueType(value) != JSON_NULL)
    {
        if (JsonValueType(value) == JSON_OBJECT)
        {
            if (!ConfigParseRunAs(lc, tConfig, JsonValueAsObject(value)))
            {
                goto error;
            }
        }
        else
        {
            Log(lc, LOG_ERR, "Config directive 'runAs' should be a JSON object");
            Log(lc, LOG_ERR, "that contains a 'uid' and 'gid'.");
            goto error;
        }
    }

    CONFIG_REQUIRE("dataDir", JSON_STRING);
    CONFIG_COPY_STRING(tConfig->dataDir);

    CONFIG_OPTIONAL_INTEGER(tConfig->threads, "threads", 1);
    CONFIG_OPTIONAL_INTEGER(tConfig->maxConnections, "maxConnections", 32);
    CONFIG_OPTIONAL_INTEGER(tConfig->maxCache, "maxCache", DB_MIN_CACHE);

    CONFIG_REQUIRE("federation", JSON_BOOLEAN);
    if (JsonValueAsBoolean(value))
    {
        tConfig->flags |= TELODENDRIA_FEDERATION;
    }

    CONFIG_REQUIRE("registration", JSON_BOOLEAN);
    if (JsonValueAsBoolean(value))
    {
        tConfig->flags |= TELODENDRIA_REGISTRATION;
    }

    CONFIG_REQUIRE("log", JSON_OBJECT);
    if (!ConfigParseLog(lc, tConfig, JsonValueAsObject(value)))
    {
        goto error;
    }

    return tConfig;

error:
    TelodendriaConfigFree(tConfig);
    return NULL;
}

void
TelodendriaConfigFree(TelodendriaConfig * tConfig)
{
    if (!tConfig)
    {
        return;
    }

    Free(tConfig->serverName);
    Free(tConfig->baseUrl);
    Free(tConfig->identityServer);

    Free(tConfig->uid);
    Free(tConfig->gid);
    Free(tConfig->dataDir);

    Free(tConfig);
}
